refactored/src/edx.rs

```diff
@@ -0,0 +1,75 @@
+use std::collections::{HashMap, HashSet};
+
+use zeroize::ZeroizeOnDrop;
+
+pub trait Edx<
+    const KEY_LENGTH: usize,
+    const TOKEN_LENGTH: usize,
+    const VALUE_LENGTH: usize,
+    Error: std::error::Error,
+>
+{
+    /// Cryptographically secure key.
+    // TODO (TBZ): Should the same type be used by both KMAC and AEAD keys?
+    type Key: ZeroizeOnDrop;
+
+    /// Cryptographically secure token used to index values inside the encrypted
+    /// dictionary.
+    type Token = [u8; TOKEN_LENGTH];
+
+    /// Fixed length value stored inside the dictionary.
+    type Value;
+
+    /// Fixed length encrypted value stored inside the encrypted dictionary.
+    type EncryptedValue;
+
+    /// Deterministically derives a cryptographic key from the given seed.
+    fn derive_keys(seed: &[u8]) -> Self::Key;
+
+    /// Deterministically transforms the given tag into a token using the given
+    /// key. The token should leak no information about the tag.
+    fn tokenize(k: &Self::Key, tag: &[u8]) -> Self::Token;
+
+    /// Queries the encrypted dictionary for the given tokens and returns the
+    /// encrypted values.
+    ///
+    /// One network call should be made for all the tokens in order to improve
+    /// the performances.
+    fn get(
+        &self,
+        tokens: HashSet<Self::Token>,
+    ) -> Result<HashMap<Self::Token, Self::EncryptedValue>, Error>;
+
+    /// Decrypts the given value with the given key.
+    fn resolve(k: &Self::Key, encrypted_value: Self::EncryptedValue) -> Result<Self::Value, Error>;
+
+    /// Encrypts the given new values and conditionally upsert them into the
+    /// encrypted dictionary.
+    ///
+    /// For each new values:
+    /// - if there is no value in both the set of old values and the encrypted
+    ///   dictionary, inserts the new value;
+    /// - if there is a value in the set of old values that matches the
+    ///   value in the encrypted dictionary, replaces it by the new value;
+    /// - if there is a value in the set of old values that does not match the
+    ///   value in the encrypted dictionary, returns the encrypted value stored;
+    /// - if there is a value in the set of old values but no value in the
+    ///   encrypted dictionary, returns an error.
+    ///
+    /// All modifications to the encrypted dictionary are *atomic*.
```

Comment:
Maybe explain that it's not the complete `upsert` that is atomic (if we reject some values because the old values don't match we can still insert other values from the same `upsert`)

Reply:
Actually the entire `upsert` should be atomic if the following property I've added is respected:
```
    /// # Error
    ///
    /// An error should be return when a value is being upserted for a token for
    /// which the EDX contains no value but an old value is passed. If this
    /// happens, no value should be upserted at all.
```

It is unclear to me what is the desired behavior in this case. This is not supposed to be a possible case, hence the returned error.

Reply:
But if an `old_value` doesn't match what is inside the database, other values (where `old_value` match) will be updated?

Reply:
We may not want to update a corrupted index?

Reply:
This is actually a much more difficult problem than I thought (cf [this document](https://www.notion.so/cosmian/Findex-regeneration-and-integrity-properties-fb4371c5171a4c949a2c59675347e8c6)). This won't be solve for this PR. I propose to raise an issue and solve this on a later time.